projects/poirazi_neuron_model/run_tm_dim_experiment.py

```diff
@@ -0,0 +1,155 @@
+# Numenta Platform for Intelligent Computing (NuPIC)
+# Copyright (C) 2017, Numenta, Inc.  Unless you have an agreement
+# with Numenta, Inc., for a separate license for this software code, the
+# following terms and conditions apply:
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+# See the GNU Affero Public License for more details.
+#
+# You should have received a copy of the GNU Affero Public License
+# along with this program.  If not, see http://www.gnu.org/licenses.
+#
+# http://numenta.org/licenses/
+# ----------------------------------------------------------------------
+
+import numpy
+import copy
+from itertools import izip as zip, count
+from nupic.bindings.algorithms import TemporalMemory as TM
+from scipy.spatial.distance import cosine
+from data_tools import generate_evenly_distributed_data_sparse
+
+
+def convert_cell_lists_to_dense(dim, cell_list, add_1 = False):
+  if add_1:
+    dense_cell_list = numpy.zeros((len(cell_list), dim + 1))
+  else:
+    dense_cell_list = numpy.zeros((len(cell_list), dim))
+  for i, datapoint in enumerate(cell_list):
+    for cell in datapoint:
+      dense_cell_list[i, int(cell)] = 1
+    if add_1:
+      dense_cell_list[i, dim] = 1
+
+  return dense_cell_list
+
+
+def run_tm_dim_experiment(test_dims = range(1400, 3100, 100),
+                          cellsPerColumn=1,
+                          num_active = 40,
+                          activationThreshold=5,
+                          initialPermanence=0.8,
+                          connectedPermanence=0.50,
+                          minThreshold=5,
+                          maxNewSynapseCount=20,
+                          permanenceIncrement=0.05,
+                          permanenceDecrement=0.00,
+                          predictedSegmentDecrement=0.000,
+                          maxSegmentsPerCell=255,
+                          maxSynapsesPerSegment=255,
+                          seed=42,
+                          num_samples = 1000,
+                          sequence_length = 20,
+                          training_iters = 1,
+                          automatic_threshold = True,
+                          save_results = True):
+  """
+  Run an experiment tracking the performance of the temporal memory given
+  different input dimensions.  The number of active cells is kept fixed, so we
+  are in effect varying the sparsity of the input.   We track performance by
+  comparing the cells predicted to be active with the cells actually active in
+  the sequence without noise at every timestep, and averaging across timesteps.
+  Three metrics are used, correlation (Pearson's r, by numpy.corrcoef),
```

Comment:
I also tested that metric (it is actually still gathered in the experiments, under the name similarity, using a simple Jaccard index calculation), but the issue is that it does not necessarily hold up well in conditions of low sparsity.  The number of overlapping bits can be quite high in those cases without there actually being any correlation whatsoever between what the system is predicting and what the true answer is.